soc: nxp: ke1xz: Enhance power management with STOP modes

Holt-Sun opened the pull request:
Implement proper STOP mode support with partial stop options, add power state timing characteristics, and enable LPTMR-based wake-up for KE1xZ SoC family.

- Fix STOP mode entry with correct SLEEPDEEP and PSTOPO configuration
- Add min-residency-us/exit-latency-us to all power states
- Use LPTMR as cortex-m-idle-timer for low-power operation
- Add power_mgmt_soc test support for FRDM-KE17Z

Depends on following PRs:
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/101392
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/101573
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/101705

ZhaoxiangJin commented on tests/subsys/pm/power_mgmt_soc/boards/frdm_ke17z.conf:
```diff
@@ -0,0 +1,8 @@
+
+# Keep LPTMR as wake/idle timer via counter API
+CONFIG_COUNTER=y
+CONFIG_COUNTER_MCUX_LPTMR=y
```
`CONFIG_COUNTER_MCUX_LPTMR `not needed manually set to y here guess.

ZhaoxiangJin commented on soc/nxp/kinetis/ke1xz/Kconfig.defconfig:
```diff
@@ -1,16 +1,13 @@
-config MCUX_LPTMR_TIMER
-	default y if PM
-
 config CORTEX_M_SYSTICK
```
Maybe we can change it to `configdefault CORTEX_M_SYSTICK` at the same time?

Holt-Sun commented:
@Albort12138 @ZhaoxiangJin , please help review.

bjarki-andreasen commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -22,22 +23,48 @@ __ramfunc static void wait_for_flash_prefetch_and_idle(void)
 void pm_state_set(enum pm_state state, uint8_t substate_id)
 {
+	__enable_irq();
+
```
`__enable_irq()`? usually IRQs are disabled before sleep to ensure any pending IRQ is left pending until the CPU enters sleep. Otherwise a race may occur if an interrupt which is expected to wake up the CPU is handled before the CPU enters sleep, thus the CPU enters sleep to never wake up again.

ZhaoxiangJin replied:
If IRQ is not enabled, the wake-up source may not be able to wake up the MCU properly. (Of course, some MCUs can be woken up as long as there is a pending interrupt.)

Holt-Sun replied:
It should be __disable_irq().  __disable_irq() won't stop wake-up source wake up the CPU. It stops irq handler with interrupts happen waking up the CPU.

Holt-Sun replied:
Here is the detailed run log with updated code:
```
pm_state_set: Entering STOP mode (substate_id=0)

[14:34:36.555]_isr_wrapper: Enter pm_system_resume in _kernel.idle.
pm_system_resume: entry
pm_system_resume: post_ops_required=1
pm_system_resume: doing post ops.
pm_state_exit_post_ops: entered.
pm_state_exit_post_ops: exited power state 2
pm_state_exit_post_ops: exiting
pm_system_resume: calling sys_clock_idle_exit.
sys_clock_idle_exit: entered
sys_clock_idle_exit: calling z_cms_lptim_hook_on_lpm_exit
z_cms_lptim_hook_on_lpm_exit: entered
z_cms_lptim_hook_on_lpm_exit: exiting
sys_clock_idle_exit: systick_us=664400 idle_timer_us=2249750
sys_clock_idle_exit: missed_cycles=76096800 dticks=23410
sys_clock_idle_exit: exiting
_isr_wrapper: Back from pm_system_resume in _kernel.idle.
_isr_wrapper: calling irq_number=29
Run mcux_lptmr_isr.
pm_state_set: Woke from STOP mode
pm_state_set: exiting
pm_system_resume: entry
pm_system_resume: post_ops_required=0
I: Wake from Deep Sleep
```

bjarki-andreasen reviewed (changes requested):
Looks good, but I'm suspicious of the `__enable_irq()`

Albort12138 commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -22,22 +23,48 @@ __ramfunc static void wait_for_flash_prefetch_and_idle(void)
 
 void pm_state_set(enum pm_state state, uint8_t substate_id)
 {
+	__enable_irq();
```
```suggestion
	__disable_irq();
```

Albort12138 commented on dts/arm/nxp/nxp_ke1xz.dtsi:
```suggestion
			cpu-power-states = <&idle &pstop2 &pstop1 &stop>;
```
Based on this order, to find the appropriate mode more efficiently

Albort12138 commented on dts/arm/nxp/nxp_ke1xz.dtsi:
```diff
@@ -33,24 +33,33 @@
 				compatible = "zephyr,power-state";
 				power-state-name = "suspend-to-idle";
 				substate-id = <1>;
+				min-residency-us = <60>;
```
Where residency come from?

ZhaoxiangJin commented on soc/nxp/kinetis/ke1xz/soc.c:
```diff
@@ -186,6 +187,9 @@ void soc_early_init_hook(void)
 	/* Initialize system clocks and PLL */
 	clk_init();
+#ifdef CONFIG_PM
+	SMC_SetPowerModeProtection(SMC, kSMC_AllowPowerModeAll);
```
I'd like to know why allow enter low power mode here?

Holt-Sun commented:
> To be honest, I'm not sure if your MCU has actually gone to low power mode. Look [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/pm.c#L207): LPTMR can only be correctly configured as the wake-up source when `exit_latency_ticks` is greater than 0. And looking at the power mode nodes in the devicetree, their `min-residency-us` values ​​are very low, possibly less than one kernel tick under the default transition strategy.

Your concern is valid. Previous systick pending interrupt forbid the CPU to stop down. Already update the code clear pending interrupt before entering stop. Now it is ok.
Here is more detailed log:
```
I: About to enter deep Sleep
PM >
LPM_ENTRY: max_lpm_time_us=1099600 ticks=4398
From isr wrapper.
In pm_system_resume: calling sys_clock_idle_exit.
LPM_EXIT: idle_timer_pre=0 idle_timer_post=4432
LPM_EXIT: pre=0 post=4432 pending=1 wrap=1 diff=8831 us=2207750
LPM_EXIT: systick_us=13205 idle_timer_us=2207750
LPM_EXIT: missed_cycles=105338160 dticks=22567
ISR Wrapper: irq_number=29
In mcux_lptmr_isr.
In pm_system_resume.
I: Wake from Deep Sleep
PM <
I: PM sleep residency 2.261 seconds
I: PM dummy single-thread completed
```

ZhaoxiangJin commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -12,32 +12,86 @@
 {
+	/* Set PRIMASK */
+	__disable_irq();
+	/* Set BASEPRI to 0 */
```
If I remember correctly, the KE1xZ ([CM0P](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/kinetis/ke1xz/Kconfig#L9)) has no BASEPRI ?

Holt-Sun commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -12,32 +12,86 @@
+		SysTick->CTRL = syst_ctrl_saved & ~SysTick_CTRL_TICKINT_Msk;
+
+		/* Clear pending core exceptions that can cause immediate WFI return */
+		SCB->ICSR = SCB_ICSR_PENDSTCLR_Msk | SCB_ICSR_PENDSVCLR_Msk;
```
1. Clear PendSV is overhead, removed.
2. The systick pending state does prevent the system enter low power mode for KE1x device with no wake up management.
3. [CORTEX_M_SYSTICK_RESET_BY_LPM](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/Kconfig.cortex_m_systick#L99) is for the case that systick is reset during the some low power modes in some SOC, in KE1X, the systick continuously run in stop mode which is not suitable with [CORTEX_M_SYSTICK_RESET_BY_LPM](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/Kconfig.cortex_m_systick#L99). Introduce [CORTEX_M_SYSTICK_STOP_IN_LPM](https://github.com/nxp-upstream/zephyr/blob/c2972636c4daa1c040c9fbd7e8045105ad84c262/drivers/timer/Kconfig.cortex_m_systick#L115) to support such case: intentionally stop systick during low power enter process.

Code has been updated, please review again.

ZhaoxiangJin commented on tests/subsys/pm/power_mgmt_soc/boards/frdm_ke17z.overlay:
```diff
@@ -0,0 +1,77 @@
+
+/* Disable other potential wake sources/peripherals */
+&adc0 {
+	status = "disabled";
```
This is not a recommend way, you should use [power-domain-soc-state-change](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/power-domain/power-domain-soc-state-change.yaml) to turn-on and turn-off peripherals. for example: [MCXW23](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_mcxw23x_common.dtsi#L73) and [RW61x ](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_rw6xx_common.dtsi#L87)

ZhaoxiangJin commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -12,35 +12,63 @@
+
+#if (defined(FSL_FEATURE_SMC_HAS_PSTOPO) && FSL_FEATURE_SMC_HAS_PSTOPO)
+		SMC->STOPCTRL = (SMC->STOPCTRL & ~SMC_STOPCTRL_PSTOPO_MASK) |
+				SMC_STOPCTRL_PSTOPO(substate_id);
```
should be aligned with the indentation of the previous row.

Holt-Sun replied:
I used clang-format to format this file, it is this result.

ZhaoxiangJin commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -12,35 +12,63 @@
 		if (SMC->PMCTRL & SMC_PMCTRL_STOPA_MASK) {
 			LOG_DBG("partial stop aborted");
 		}
+		irq_unlock(0);
```
same as above

ZhaoxiangJin reviewed (commented):
we should split the PR into different commits

Holt-Sun commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -12,35 +12,63 @@
+		__DSB();
+		__ISB();
+		__WFI();
+		irq_unlock(0);
```
`I have a question to this log, the sleep time is 2249750us in STOP mode (SoC idle mode)? In the device tree, the min-residency-us of idle mode is 10.` This is a good catch. The issue may not be in this PR, but in https://github.com/zephyrproject-rtos/zephyr/blob/cdb7debfb726569440fc22ce6673de2b86c728f7/drivers/timer/cortex_m_systick.c#L176-L208 : 
lin179 gets the idle_timer_post which is same as idle_timer_scheduled_sleep_ticks , line181 gets the top, actually, they are same for most cases(lptmr frequency is usually much lower than CPU frequency), in line 197-199, the idle_timer_int_pending is true(which may be an issue), then it is always wrap. so the calculated the sleep time is wrong, if we fix it, then the sleep time is normal:
`sys_clock_idle_exit: systick_us=0 idle_timer_us=1099500 systick_diff=4294967296 missed_cycles=52776000 dticks=11689`
I would think it is more like issue in [cortex_m_systick.c](https://github.com/zephyrproject-rtos/zephyr/blob/cdb7debfb726569440fc22ce6673de2b86c728f7/drivers/timer/cortex_m_systick.c).

If you see scenario 2, the idle time is expected which I get counter from [idle_timer_alarm_stub](https://github.com/zephyrproject-rtos/zephyr/blob/cdb7debfb726569440fc22ce6673de2b86c728f7/drivers/timer/cortex_m_systick.c#L126). I will try to fix this in PR https://github.com/zephyrproject-rtos/zephyr/pull/101705

`From the log, it appears there was a premature wakeup. The irq_unlock(0); occurred after WFI (Wait For Interrupt). I don't understand why this situation would happen. This is worth a deeper analysis, but for now, let's leave it as is.`
The irq_unlock(0) occurred after WFI, but not immediately, the LPTMR interrupt(no matter irq lock or not) wakes up the CPU when set alarm time hits, the PC will run irq_unlock(0) to open the irq, then PC will run into [_isr_wrapper](https://github.com/zephyrproject-rtos/zephyr/blob/cdb7debfb726569440fc22ce6673de2b86c728f7/arch/arm/core/cortex_m/isr_wrapper.c#L33), the time duration is idle_timer_pre + idle_timer_post = 4600. There is also a "[14:31:20.872]" to tip that there is time passed. So there is no prematrue wakeup here.

Holt-Sun commented on soc/nxp/kinetis/ke1xz/Kconfig.defconfig:
```diff
@@ -1,17 +1,14 @@
 if SOC_SERIES_KE1XZ
 
-config MCUX_LPTMR_TIMER
+configdefault CORTEX_M_SYSTICK
```
Removed.

Holt-Sun committed:
soc: nxp: ke1xz: Add power management support

Implement power management with IDLE, STOP, PSTOP1, and PSTOP2 modes.

- Add power state definitions with timing parameters
- Implement pm_state_set() with proper SLEEPDEEP handling
- Add XIP-safe WFI execution from RAM
- Enable SMC driver and power mode protection
- Remove forced timer Kconfig defaults

Signed-off-by: Holt Sun <holt.sun@nxp.com>

Holt-Sun committed:
boards: frdm_ke17z: Add power management support

Enable power management using LPTMR0 as the idle timer.

- Set zephyr,cortex-m-idle-timer to lptmr0
- Remove redundant stop state DTS overrides
- Add power_mgmt_soc test configuration and overlay

Signed-off-by: Holt Sun <holt.sun@nxp.com>

Holt-Sun commented:
@bjarki-andreasen , all dependencies and issues have been fixed, please review. Thank you.
